libraries/AP_Frsky_Telem/frsky.cpp

```diff
@@ -0,0 +1,220 @@
+#include <AP_HAL/AP_HAL.h>
+#include "frsky.h"
+
+/*
+   State machine to process incoming Frsky SPort bytes
+*/
+bool FRSky::parse_sport_telemetry_data(uint8_t data, sport_parse_state_t &parse_state)
+{
+    switch (parse_state.state) {
+        case STATE_DATA_START:
+            if (data == FRAME_HEAD) {
+                parse_state.state = STATE_DATA_IN_FRAME ;
+                parse_state.telemetry_rx_buffer_count = 0;
+            } else {
+                if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                    parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data;
+                }
+                parse_state.state = STATE_DATA_IN_FRAME;
+            }
+            break;
+
+        case STATE_DATA_IN_FRAME:
+            if (data == FRAME_DLE) {
+                parse_state.state = STATE_DATA_XOR; // XOR next byte
+            } else if (data == FRAME_HEAD) {
+                parse_state.state = STATE_DATA_IN_FRAME ;
+                parse_state.telemetry_rx_buffer_count = 0;
+                break;
+            } else if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data;
+            }
+            break;
+
+        case STATE_DATA_XOR:
+            if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data ^ STUFF_MASK;
+            }
+            parse_state.state = STATE_DATA_IN_FRAME;
+            break;
+
+        case STATE_DATA_IDLE:
+            if (data == FRAME_HEAD) {
```

Comment:
I ported the parser from OpenTX, I need draw down the state machine and double check, right now it "just works"

Reply:
I'd suggest we have enough of our home-grown bugs to work through; importing new ones seems strange.

Is there a reason that two consecutive FRAME_HEAD bytes?

Reply:
I assumed it was well tested. Well formed packets should never have the transition START->frame_head, I can try to simplify the state machine but rather than a bug I think the OpenTX guys hit some bad behaving tx/rx combo, I just don't wan't to reinvent the wheel like I had to do for fport "undocumented" feature vs bug packet rate control